tests/integration/test_charm.py

```diff
@@ -175,3 +181,72 @@ async def test_prometheus_grafana_integration(ops_test: OpsTest):
     wait=tenacity.wait_exponential(multiplier=1, min=1, max=10),
     reraise=True,
 )
+
+
+@pytest.mark.abort_on_fail
+async def test_remove_and_upgrade(ops_test: OpsTest):
+    """Test remove and upgrade.
+
+    This test should be last in the suite, because it removes deployed charm.
+    First, remove is tested by removing previously deployed charm and verifying that resources are
+    also removed.
+    Second, a stable version of charm is deployed and refresh (upgrade) is performed and resources
+    are verified for correct versions.
+    """
+    # remove deployed charm and verify that it is removed
+    await ops_test.model.remove_application(app_name=APP_NAME, block_until_done=True)
+    assert APP_NAME not in ops_test.model.applications
+
+    # verify that all CRDs are removed
+    lightkube_client = lightkube.Client()
+    crd_list = lightkube_client.list(
+        CustomResourceDefinition,
+        labels=[("app.juju.is/created-by", "training-operator")],
+        namespace=ops_test.model.name,
+    )
+    # testing for empty list (iterator)
+    _last = object()
+    assert next(crd_list, _last) is _last
+
+    # deploy stable version of the charm
+    await ops_test.model.deploy(entity_url=APP_NAME, channel="1.5/stable", trust=True)
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", raise_on_blocked=True, timeout=60 * 10
+    )
+
+    # refresh (upgrade) using charm built in test_build_and_deploy()
+    # NOTE: using ops_test.juju() because there is no functionality to refresh in ops_test
+    image_path = METADATA["resources"]["training-operator-image"]["upstream-source"]
+    await ops_test.juju(
+        "refresh",
+        APP_NAME,
+        f"--path={CHARM_LOCATION}",
+        f'--resource="training-operator-image={image_path}"',
+    )
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", raise_on_blocked=True, timeout=60 * 10
+    )
+
+    # verify that all CRDs are installed
```

Comment:
Right now the training-operator is also deploying auth stuff, maybe you want to check that also.

Reply:
Will do.